src/sync.h

```diff
@@ -53,8 +53,9 @@ void LeaveCritical();
 void CheckLastCritical(void* cs, std::string& lockname, const char* guardname, const char* file, int line);
 std::string LocksHeld();
 template <typename MutexType>
-void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) ASSERT_EXCLUSIVE_LOCK(cs);
-void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, void* cs);
+void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(cs);
+template <typename MutexType>
+void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(!cs);
```

Comment:
These can also be tagged with `ASSERT_EXCLUSIVE_LOCK`.

```suggestion
void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(cs);
template <typename MutexType>
void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(!cs);
```

Reply:
I tried the following with `AssertLockHeldInternal() EXCLUSIVE_LOCKS_REQUIRED(cs) ASSERT_EXCLUSIVE_LOCK(cs);`:

```cpp
void f()
{
    // The following line produces a warning due to the EXCLUSIVE_LOCKS_REQUIRED attribute:
    // Calling function 'AssertLockHeldInternal<AnnotatedMixin<std:: __1::recursive_mutex>>' requires holding mutex 'cs_main' exclusively [-Wthread-safety-analysis]
    AssertLockHeld(cs_main);

    // The following line produces a warning due to the ASSERT_EXCLUSIVE_LOCK attribute:
    // Acquiring mutex 'cs_main' that is already held
    LOCK(cs_main);
}
```

However if the assert attribute is defined first: `ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(cs)` then only the second warning is generated (`EXCLUSIVE_LOCKS_REQUIRED` is ignored in this case).

Reply:
> These can also be tagged with `ASSERT_EXCLUSIVE_LOCK`.

What are expected benefits?

Reply:
@vasild 
Here are branches to test your suggestion:
- https://github.com/hebasto/bitcoin/tree/pr16981-base
- https://github.com/hebasto/bitcoin/tree/pr16981-suggestion

Compiling the branch with suggestion miss the following warnings:
```
wallet/scriptpubkeyman.cpp:1861:9: warning: reading variable 'm_map_crypted_keys' requires holding mutex 'cs_desc_man' [-Wthread-safety-analysis]
        m_map_crypted_keys[pubkey.GetID()] = make_pair(pubkey, crypted_secret);
        ^
wallet/scriptpubkeyman.cpp:1864:9: warning: reading variable 'm_map_keys' requires holding mutex 'cs_desc_man' [-Wthread-safety-analysis]
        m_map_keys[pubkey.GetID()] = key;

```

and

```
wallet/wallet.cpp:4429:54: warning: passing variable 'vMasterKey' by reference requires holding mutex 'cs_wallet' [-Wthread-safety-reference]
                if (!spk_manager->CheckDecryptionKey(vMasterKey) && !spk_manager->Encrypt(vMasterKey, nullptr)) {
                                                     ^
wallet/wallet.cpp:4429:91: warning: passing variable 'vMasterKey' by reference requires holding mutex 'cs_wallet' [-Wthread-safety-reference]
                if (!spk_manager->CheckDecryptionKey(vMasterKey) && !spk_manager->Encrypt(vMasterKey, nullptr)) {

```
```

Reply:
Ok, I confirm the above. The presence of `ASSERT_EXCLUSIVE_LOCK()` causes some warnings to be silenced.

Reply:
> However if the assert attribute is defined first: `ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(cs)` then only the second warning is generated (`EXCLUSIVE_LOCKS_REQUIRED` is ignored in this case).

This is somewhat expected. If you assert `cs`, it's assumed to be held afterwards. If you require it to be held and then assert it, we warn if it isn't initially held, and then assume it to be held afterwards.

Generally though you shouldn't expect attributes to be treated in the order they are given, and this particular combination of attributes just doesn't make sense. (Either you're assuming a capability that you've just asserted, which is redundant, or you assert a capability that you've previously assumed, which is pointless.)

Reply:
@aaronpuchert,
> This is somewhat expected. If you assert cs, it's assumed to be held afterwards. If you require it to be held and then assert it, we warn if it isn't initially held, and then assume it to be held afterwards.

I agree with that if they were two distinct functions:
```cpp
func_tagged_with_assert_exclusive_lock();
...
func_tagged_with_exclusive_lock_required();
...
```

But for a single function: `void f() ATTR1 ATTR2` I wouldn't expect the behavior to differ if defined like `void f() ATTR2 ATTR1` instead.

If the combination of `ATTR1` and `ATTR2` is meaningless then maybe it would be good to produce some warning or error instead of silently assuming some behavior. Like for `signed unsigned int x;`.

Reply:
> If the combination of `ATTR1` and `ATTR2` is meaningless then maybe it would be good to produce some warning or error instead of silently assuming some behavior.

Makes sense, but there are many attributes and I'm not sure whether we can exhaustively figure out which combinations make sense and which don't.